tasks/cephfs/vstart_runner.py

```diff
@@ -0,0 +1,782 @@
+            def __init__(self, mount_daemon):
+                self.mount_daemon = mount_daemon
+
+            def close(self):
+                self.mount_daemon.kill()
+
+        return FakeStdIn(self)
+
+
+class LocalRemote(object):
+    """
+    Amusingly named class to present the teuthology RemoteProcess interface when we are really
+    running things locally for vstart
+
+    Run this inside your src/ dir!
+    """
+
+    def __init__(self):
+        self.name = "local"
+        self.hostname = "localhost"
+
+    def run(self, args, check_status=True, wait=True, stdout=None, stderr=None, cwd=None, stdin=None):
+        # We don't need no stinkin' sudo
+        args = [a for a in args if a != "sudo"]
+
+        for arg in args:
+            if not isinstance(arg, basestring):
+                raise RuntimeError("Oops, can't handle arg {0} type {1}".format(
+                    arg, arg.__class__
+                ))
+
+        log.info("Running {0}".format(args))
+        subproc = subprocess.Popen(args,
+                                   stdout=subprocess.PIPE,
+                                   stderr=subprocess.PIPE,
+                                   stdin=subprocess.PIPE,
+                                   cwd=cwd)
+
+        if stdin and isinstance(stdin, basestring):
+            # Hack: writing to stdin is not deadlock-safe, but it "always" works
+            # as long as the input buffer is "small"
+            subproc.stdin.write(stdin)
+
+        proc = LocalRemoteProcess(
+            args, subproc, check_status,
+            stdout, stderr
+        )
+
+        if wait:
+            proc.wait()
+
+        return proc
+
+
+# FIXME: twiddling vstart daemons is likely to be unreliable, we should probably just let vstart
+# run RADOS and run the MDS daemons directly from the test runner
+class LocalDaemon(object):
+    def __init__(self, daemon_type, daemon_id):
+        self.daemon_type = daemon_type
+        self.daemon_id = daemon_id
+        self.controller = LocalRemote()
+
+    @property
+    def remote(self):
+        return LocalRemote()
+
+    def running(self):
+        return self._get_pid() is not None
+
+    def _get_pid(self):
+        """
+        Return PID as an integer or None if not found
+        """
+        ps_txt = self.controller.run(
+            args=["ps", "aux"]
+        ).stdout.getvalue().strip()
+        lines = ps_txt.split("\n")[1:]
+
+        for line in lines:
+            if line.find("ceph-{0} -i {1}".format(self.daemon_type, self.daemon_id)) != -1:
+                log.info("Found ps line for daemon: {0}".format(line))
+                return int(line.split()[1])
+
+        return None
+
+    def wait(self, timeout):
+        waited = 0
+        while self._get_pid() is not None:
+            if waited > timeout:
+                raise MaxWhileTries("Timed out waiting for daemon {0}.{1}".format(self.daemon_type, self.daemon_id))
+            time.sleep(1)
+            waited += 1
```

Comment:
Should probably use a start time and actually look at the clock instead of relying on an increment...it's unlikely to be a serious problem since this is just a timeout but they get inaccurate surprisingly quickly on many systems.